### PR TITLE
[MS-223] CLI Filtering

### PR DESCRIPTION
**Pull request by imda-kelvinkok**
1. start the cli with python -m moonshot cli interactive
2. once inside, you can test all the list with the find command: i.e. list_cookbooks -f "risk" . this will return all cookbooks with the word risk in any of its attribute (i.e. if the cookbook's ID has the word risk or if one of the recipes has the word risk). you can test the following command with the -f "your keyword" :
- list_cookbooks
- list_datasets
- list_metrics
- list_recipes
- list_results
- list_runs
- list_endpoints
- list_connector_types
- list_prompt_templates
- list_attack_modules
- list_context_strategies
- list_sessions

### CODE REVIEW

**Comment by imda-kelvinkok** [comment]
image

**Comment by imda-kelvinkok** [comment]
image

Unit tests for:
- list_attack_modules
- list_context_strategies
- list_sessions

**Comment by imda-kelvinkok** [comment]
image

Unit tests for:
- list_connectors
- list_endpoints
- list_prompt_templates

**Review by imda-normanchia — APPROVED** [review]

File: moonshot/integrations/cli/benchmark/recipe.py [hunk 1]
```diff
@@ -382,9 +398,9 @@ def show_recipe_results(recipes, endpoints, recipe_results, duration):
         console.print("[red]There are no results.[/red]")
 
     # Print run stats
-    console.print(
-        f"{'='*50}\n[blue]Time taken to run: {duration}s[/blue]\n*Overall rating will be the lowest grade that the recipes have in each cookbook\n{'='*50}"
-    )
+    run_stats = f"""{'='*50}\n[blue]Time taken to run: {duration}s[/blue]\n*Overall rating will be the lowest grade
```
Comment (imda-lionelteo) [inline moonshot/integrations/cli/benchmark/recipe.py:401]: do we really need this?

File: moonshot/integrations/cli/benchmark/cookbook.py [hunk 2]
```diff
@@ -340,9 +356,9 @@ def show_cookbook_results(cookbooks, endpoints, cookbook_results, duration):
         console.print("[red]There are no results.[/red]")
 
     # Print run stats
-    console.print(
-        f"{'='*50}\n[blue]Time taken to run: {duration}s[/blue]\n*Overall rating will be the lowest grade that the recipes have in each cookbook\n{'='*50}"
-    )
+    run_stats = f"""{'='*50}\n[blue]Time taken to run: {duration}s[/blue]\n*Overall rating will be the lowest grade
```
Comment (imda-lionelteo) [inline moonshot/integrations/cli/benchmark/cookbook.py:359]: do we really need this

File: moonshot/integrations/cli/common/prompt_template.py [hunk 3]
```diff
@@ -3,22 +3,44 @@
...
+                return filtered_prompt_templates_list
+            else:
+                print("No prompt templates containing keyword found.")
+                return None
+        else:
+            display_prompt_templates(prompt_templates_list)
+            return prompt_templates_list
     except Exception as e:
         print(f"[list_prompt_templates]: {str(e)}")
+        return None
```
Comment (imda-lionelteo) [inline moonshot/integrations/cli/common/prompt_template.py:43]: do we need to return None here?

File: moonshot/integrations/cli/redteam/attack_module.py [hunk 4]
```diff
@@ -3,38 +3,42 @@
...
-            for k, v in attack_module_data.items():
-                attack_module_data_str += f"[blue]{k.capitalize()}:[/blue] {v}\n\n"
-            table.add_row(str(attack_module_index), attack_module_data_str)
+    Returns:
+         list | None: A list of AttackModule or None if there is no result.
+    """
+    try:
+        print("Listing attack modules may take a while...")
+        attack_module_metadata_list = api_get_all_attack_module_metadata()
+        keyword = args.find.lower() if args.find else ""
```
Comment (imda-lionelteo) [inline moonshot/integrations/cli/redteam/attack_module.py:26]: remove this empty sentence

File: moonshot/integrations/cli/redteam/context_strategy.py [hunk 5]
```diff
@@ -52,30 +53,36 @@ def use_context_strategy(args: argparse.Namespace) -> None:
...
-            for k, v in context_strategy_data.items():
-                context_strategy_data_str += f"[blue]{k.capitalize()}:[/blue] {v}\n\n"
-            table.add_row(str(context_strategy_index), context_strategy_data_str)
-        console.print(table)
-    else:
-        console.print("[red]There are no context strategies found.[/red]", style="bold")
+    try:
+        context_strategy_metadata_list = api_get_all_context_strategy_metadata()
+        keyword = args.find.lower() if args.find else ""
+
```
Comment (imda-lionelteo) [inline moonshot/integrations/cli/redteam/context_strategy.py:70]: remove this empty sentence

File: moonshot/integrations/cli/redteam/session.py [hunk 6]
```diff
@@ -128,33 +129,40 @@ def end_session() -> None:
...
-    if session_metadata_list:
-        table = Table(
-            title="Session List", show_lines=True, expand=True, header_style="bold"
-        )
-        table.add_column("No.", justify="left", width=2)
-        table.add_column("Session ID", justify="left", width=20)
-        table.add_column("Contains", justify="left", width=78)
+    try:
+        session_metadata_list = api_get_all_session_metadata()
+        keyword = args.find.lower() if args.find else ""
```
Comment (imda-lionelteo) [inline moonshot/integrations/cli/redteam/session.py:149]: remove this empty space

File: moonshot/integrations/cli/redteam/session.py [hunk 7]
```diff
@@ -128,33 +129,40 @@ def end_session() -> None:
...
+            )
+            if filtered_session_metadata_list:
+                display_sessions(filtered_session_metadata_list)
+                return filtered_session_metadata_list
+            else:
+                print("No sessions containing keyword found.")
+                return None
+        else:
+            display_sessions(session_metadata_list)
+            return session_metadata_list
```
Comment (imda-lionelteo) [inline moonshot/integrations/cli/redteam/session.py:163]: remove this empty space

File: moonshot/integrations/cli/redteam/session.py [hunk 8]
```diff
@@ -351,20 +359,39 @@ def delete_bookmark(args) -> None:
...
 
     Returns:
-        None
+        list | None: A list of Bookmark or None if there is no result.
     """
     try:
         bookmarks_list = api_get_all_bookmarks()
-        display_bookmarks(bookmarks_list)
+        keyword = args.find.lower() if args.find else ""
+
```
Comment (imda-lionelteo) [inline moonshot/integrations/cli/redteam/session.py:381]: remove this empty space

File: moonshot/integrations/cli/redteam/session.py [hunk 9]
```diff
@@ -351,20 +359,39 @@ def delete_bookmark(args) -> None:
...
+            if filtered_bookmarks_list:
+                display_bookmarks(filtered_bookmarks_list)
+                return filtered_bookmarks_list
+            else:
+                print("No bookmarks containing keyword found.")
+                return None
+        else:
+            display_bookmarks(bookmarks_list)
+            return bookmarks_list
+
```
Comment (imda-lionelteo) [inline moonshot/integrations/cli/redteam/session.py:394]: remove this empty space

File: moonshot/integrations/cli/redteam/session.py [hunk 10]
```diff
@@ -351,20 +359,39 @@ def delete_bookmark(args) -> None:
...
+        list | None: A list of Bookmark or None if there is no result.
     """
     try:
         bookmarks_list = api_get_all_bookmarks()
-        display_bookmarks(bookmarks_list)
+        keyword = args.find.lower() if args.find else ""
+
+        if keyword:
+            filtered_bookmarks_list = find_keyword(keyword, bookmarks_list)
+
```
Comment (imda-lionelteo) [inline moonshot/integrations/cli/redteam/session.py:384]: remove this empty space

File: moonshot/integrations/cli/benchmark/recipe.py (hunk 1 above)
Comment (imda-kelvinkok) [reply moonshot/integrations/cli/benchmark/recipe.py:401]: added a variable so it is easier to format the print statement and break them into smaller chunks

File: moonshot/integrations/cli/benchmark/cookbook.py (hunk 2 above)
Comment (imda-kelvinkok) [reply moonshot/integrations/cli/benchmark/cookbook.py:359]: added a variable so it is easier to format the print statement and break them into smaller chunks

File: moonshot/integrations/cli/common/prompt_template.py (hunk 3 above)
Comment (imda-kelvinkok) [reply moonshot/integrations/cli/common/prompt_template.py]: removed this

File: moonshot/integrations/cli/redteam/session.py (hunk 10 above)
Comment (imda-kelvinkok) [reply moonshot/integrations/cli/redteam/session.py]: removed

File: moonshot/integrations/cli/redteam/session.py (hunk 9 above)
Comment (imda-kelvinkok) [reply moonshot/integrations/cli/redteam/session.py]: removed

File: moonshot/integrations/cli/redteam/session.py (hunk 8 above)
Comment (imda-kelvinkok) [reply moonshot/integrations/cli/redteam/session.py]: removed

File: moonshot/integrations/cli/redteam/session.py (hunk 7 above)
Comment (imda-kelvinkok) [reply moonshot/integrations/cli/redteam/session.py]: removed

File: moonshot/integrations/cli/redteam/session.py (hunk 6 above)
Comment (imda-kelvinkok) [reply moonshot/integrations/cli/redteam/session.py]: removed

File: moonshot/integrations/cli/redteam/context_strategy.py (hunk 5 above)
Comment (imda-kelvinkok) [reply moonshot/integrations/cli/redteam/context_strategy.py]: removed

File: moonshot/integrations/cli/redteam/attack_module.py (hunk 4 above)
Comment (imda-kelvinkok) [reply moonshot/integrations/cli/redteam/attack_module.py]: removed

**Review by imda-lionelteo — APPROVED** [review]

**Pull request merged**